### PR TITLE
【Task.31】feat(sft): add pair-aware DPO training

### PR DESCRIPTION
**Pull request by A-Words**
## What

本 PR 在现有离线 SFT 链路中加入 pair-aware Direct Preference Optimization（DPO）能力。

- 增加严格的 chosen/rejected 偏好数据 schema 校验、渲染、截断、拒绝原因统计，以及确定性的 UltraFeedback 子集生成。
- 在 TransferQueue、`SeqlenBalancedSampler`、DP 分配、动态 batching 和 micro-batch packing 全链路保持 preference pair 原子性。
- 增加标准 reference DPO、reference-free DPO loss、pair-count 归一化和 DPO 指标。
- 增加事务化 frozen-reference 重建，并校验 canonical parameter、源文件 manifest、optimizer state 和固定 probe digest。
- 将验证后的 reference identity 写入 checkpoint sidecar，并在 fresh/resume 训练中持续校验。
- 增加公开的 Qwen3-0.6B/UltraFeedback recipe、中英文文档和聚焦测试。

本 PR 不增加 Reward Model scalar head。Reward Modeling 和最终的共享偏好评测器位于后续 PR #258。

## Why

Relax 当前主要覆盖在线 RL 工作流。Task 31 要求在不复制训练框架的前提下，复用 SFT trainer 建立标准的离线偏好优化链路。

实现遵循 [RFC #208](https://github.com/redai-studio/Relax/issues/208) 和[Task 31 要求](https://github.com/redai-infra/community/blob/main/contributor-program/2026-cohort-1/official-task.md#31-dpo--reward-modeling高级)。

Relates to #208.

Follow-up: #258.

## How

本次更新合并主线 `0651812093e3cd730302709b3db35ce0bb199e4b`，保留作者历史；主线集成提交为 [`14ccc2a`](https://github.com/redai-studio/Relax/commit/14ccc2a)。

- 通过现有 `_create_sft_train_dataset` / `_prepare_sft_tq_payload` 接入偏好数据，保留主线 SFT producer/shard/prepack 路径。
- preference v1 明确拒绝 `--sft-async-prepack`、`task_type=seq_cls` 和 MTP-only 组合；CPU 数据预取仍可用。
- 补齐随机 train/eval split 所需的数据集接口与恢复测试，将实际 pair 数传入 scheduler 和指标分母。#257 本身仍不启用 held-out evaluator，评估由 #258 补充。

每个 TransferQueue row 表示一个完整 preference pair，并通过 `custom_meta` 携带 chosen/rejected 的合并 sequence cost。固定版本的 `SeqlenBalancedSampler` 因而可以在不拆分 pair 的前提下形成等大小 DP 分组。动态 packer 先分配完整 pair，仅在构造模型输入时展平 chosen/rejected branch。

DPO loss 消费显式的 pair/branch identity，使用 completion-only mask，按全局 pair 数归一化，同时支持标准 reference 和 reference-free 公式。空 mask、非有限值、prompt prefix 不一致、截断后两侧相同、超大 pair，以及不一致的 global pair count 均 fail closed。

标准 DPO fresh 训练将初始 actor 快照为 frozen reference；resume 时则从固定 HF repository/revision 执行 model-only reference 重建。Canonical parameter、源文件 manifest、optimizer master/state 和固定 log-prob probe digest 可防止 reference 切换静默修改 actor 或 optimizer。验证后的 identity 会写入 checkpoint 旁的 `relax_dpo_reference.json`。

### 删除源码字符串断言（`aab0c72`）

- 删除 `test_preference_iterator_has_no_device_scalar_readback` 及其 `inspect` 导入，保留 pair batching、DP 分配和分母校验的行为测试。
- 全量 pre-commit 通过；本次仅删除测试，未在本地重跑需 Megatron/CUDA 的 GPU 测试，新提交的 CI 结果以 checks 为准。

### CI 测试修正（`4bc9dda`）

- 已回退 CPU CI stub 特判提交，删除 `test_pinned_seqlen_sampler_consumes_pair_costs_and_keeps_equal_dp_groups` 及固定源码 SHA-256 检查。
- missing-media 测试在启动后台 prefetch 前安装 warning 捕获，消除捕获顺序竞态。
- 使用当前 CPU CI 的 TransferQueue stub 执行 preference utils 与 streaming dataset 回归：**66 passed**；全量 pre-commit 通过。
- 最新提交的 GitHub CI 结果以本 PR checks 为准。

### 主线集成提交 `14ccc2a` 的本地验证

- `pre-commit run --all-files --show-diff-on-failure`：通过。
- Python 3.12 CPU 聚焦回归：**234 passed, 7 skipped**。
- 覆盖 SFT engine / producer / sharding、偏好数据截断与 split/resume、独立数值 loss、reference identity、参数校验和 Docker 固定版本 TransferQueue sampler。
- 跳过项需要 Megatron/CUDA，包括生产后端 loss/batching 和部分 reference/训练数据路径。本次未执行完整 `pytest tests/`，未重跑 GPU 训练、DP=2 或 distributed checkpoint save/resume；最新 CI 以本 PR checks 为准。

```bash
python -m pytest tests/engine/sft/ tests/components/test_sft.py \
  tests/backends/megatron/test_preference_batching.py \
  tests/backends/megatron/test_dpo_loss.py \
  tests/backends/megatron/test_dpo_reference_integrity.py \
  tests/backends/megatron/test_sft_train_data_fields.py \
  tests/utils/training/test_preference_utils.py \
  tests/data/test_prepare_ultrafeedback_preferences.py \
  tests/utils/test_arguments_sft.py -q -ra --tb=short
```

### 作者提供的历史 GPU 验证

以下日志、曲线和证据包对应旧集成提交，保留用于追溯，**不代表本次合并后的 GPU 复测结果**。主线已更新 shuffle 行为，重新验证时应比较新版本的连续训练与恢复训练。

在最终堆叠集成 head `b57a1818f2f80d28d0ad650b314eb4d1c82df3aa` 的标准容器中执行：

```text
pytest tests/engine/sft/ \
  tests/backends/megatron/test_sft_train_data_fields.py \
  tests/utils/training/test_preference_utils.py \
  tests/backends/megatron/test_preference_batching.py \
  tests/backends/megatron/test_dpo_loss.py \
  tests/backends/megatron/test_reward_model_loss.py \
  tests/backends/megatron/test_reward_model_checkpoint.py

179 passed
```

```text
pre-commit run --all-files
All hooks passed
```

固定 TransferQueue 兼容性测试覆盖 pair-row 原子性、等大小 DP 分组、row-aligned `total_lengths` cost 消费，以及无丢失、无重复的完整分配。数据准备脚本的测试覆盖 manifest：4,096 个 train pair、512 个互不重叠的 eval pair、确定性选择、拒绝原因计数和派生文件 hash。

GPU 验证固定使用相同的模型/data revision 和 global batch size 32 pairs：

| 运行 | Steps | Final loss | Margin | Tie-aware accuracy | Bootstrap lower 95% |
|---|---:|---:|---:|---:|---:|
| DPO DP=1 | 200 | 0.630386 | 0.186692 | 0.654297 | 0.619141 |
| DPO DP=2 | 200 | 0.628974 | 0.190042 | 0.664062 | 0.628906 |

- DP=1 resume 从 iteration 199 恢复，完成下一 step，并保持相同的 reference identity。
- Reference-free DPO 完成 20 steps，没有 reference log-prob 指标、reference checkpoint load 或 reference sidecar。
- DP=2 日志明确记录 `SeqlenBalancedSampler with dp_size=2`，两张 A100 均有实际活动。

## Type of Change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Documentation update
- [ ] Refactoring (no functional changes)
- [ ] Performance improvement
- [ ] CI/CD or build changes

## Screenshots / Logs

本 PR 附带 DPO DP=1/DP=2 训练曲线和脱敏后的可复现性证据包。证据包包含：

- 200-step 训练日志、scalar CSV 和 GPU telemetry；
- step 0/final 逐 pair 评测结果与冻结 batch plan；
- DP=1 resume、reference-free DPO、probe contract 和 reference sidecar；
- 固定环境、模型/data revision、镜像身份和逐文件 SHA-256 清单。

公开附件不包含 checkpoint、optimizer state、模型权重、完整数据集、凭据或机器专属标识。

测量环境使用 `ghcr.io/redai-infra/relaxrl:dev-20260723-8cc1e8fd`；证据采集时从 GHCR 解析的 manifest digest 为 `sha256:8dc39af377a570e6cd7ec88c8b7fcd44c1eb820111e9d2069f1c7c3024b2ea23`。Transformer Engine 无法导入顶层 `flash_attn_interface`，并明确回退到 native attention。因此 DP 对比描述的是相同 native-attention 环境，不是 FlashAttention 性能结果。

[task31-pr1-dpo-evidence-public-v2.tar.gz](https://github.com/user-attachments/files/31305744/task31-pr1-dpo-evidence-public-v2.tar.gz)

task31-pr1-dpo-dp1-curves
task31-pr1-dpo-dp2-curves

### CODE REVIEW

File: tests/utils/training/test_preference_utils.py [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+def test_pinned_seqlen_sampler_consumes_pair_costs_and_keeps_equal_dp_groups():
+    """Exercise the Docker-pinned TransferQueue sampler, not a local stand-
+    in."""
+    transfer_queue = pytest.importorskip("transfer_queue")
+    sampler_type = transfer_queue.SeqlenBalancedSampler
+    source_path = Path(inspect.getsourcefile(sampler_type) or "")
+    normalized_source = source_path.read_bytes().replace(b"\r\n", b"\n")
+    assert hashlib.sha256(normalized_source).hexdigest() == (
+        "dc6c2db50df4b9448d4845ccacef67a400517db892b5cd55de2e22f6baf6888b"
+    )
```
[the remark this answers is not in the document]
Comment (SigureMo) [reply tests/utils/training/test_preference_utils.py]: ……什么测试

[the remark this answers is not in the document]
Comment (SigureMo) [reply tests/utils/training/test_preference_utils.py]: 没让你回我……笨

你要是以后敢写出来这种测试就不给你 token 了

File: tests/backends/megatron/test_preference_batching.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+        assert len(set(batch["preference_branch_pair_ids"])) == 1
+        assert set(batch["preference_is_chosen"]) == {False, True}
+        seen.extend(batch["preference_branch_pair_ids"])
+    assert sorted(seen) == [100, 100, 101, 101]
+
+
+def test_preference_iterator_has_no_device_scalar_readback():
+    source = inspect.getsource(data_module._get_preference_data_iterator)
+    assert ".item(" not in source
+    assert "all_reduce(" not in source
```
Comment (SigureMo) [inline tests/backends/megatron/test_preference_batching.py]: 啧
